docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,650 @@
+
+This is the general case of a node dying: we must always assume that the
+unresponsive node might be network partitioned and still writing to S3.
+
+#### Case A: re-attachment to other nodes
+
+1. Let's say node 0 fails in a cluster of three nodes 0, 1, 2.
+2. Some external mechanism notices that the node is unavailable and initiates
+   movement of all tenants attached to that node to a different node. In
+   this example it would mean incrementing the attachment generation
+   of all tenants that were attached to node 0, and attaching them
+   to node 1 or 2 based on some distribution rule (this might be round
+   robin, or capacity based).
+3. A tenant which is now attached to node 1 will _also_ still be attached to node
+   0, from the perspective of node 0. Let's say it was originally attached with
+   generation suffix 00000001-0000-00000001 -- its new suffix on node 1
+   might be 00000002-0001-00000001 (attachment gen 2, on node 1).
+4. S3 writes will continue from nodes 0 and 1: there will be an index_part.json-00000001-0000-00000001
+   \_and\* an index_part.json-00000002-0001-00000001. Objects written under the old suffix
+   after the new attachment was created do not matter from the rest of the system's
+   perspective: the endpoints are reading from the new attachment location. Objects
+   written by node 0 are just garbage that can be cleaned up at leisure. Node 0 will
+   not do any deletions because it can't synchronize with control plane.
+
+Eventually, node 0 will somehow realize it should stop running by some means, although
+this is not necessary for correctness:
+
+- A hard-stop of the VM it is running on
+- It tries to communicate with another peer that sends it an error response
+  indicating its generation is out of date
+- It tries to do some deletions, and discovers when synchronizing with the
+  control plane that its generation number is stale.
+
+#### Case B: direct node replacement with blank node (cold standby)
+
+1. Let's say node 0 fails, and there may be some other peers but they aren't relevant.
+2. Some external mechanism notices that the node is unavailable, and creates
+   a "new node 0" (Node 0b) which is a physically separate server. The original node 0
+   (Node 0a) may still be running.
+3. On startup, node 0b acquires a new generation number. It doesn't have any local state,
+   so the control plane must send attach requests to node 0b for all the tenants. However,
+   these do not have to increment the attachment generation as they're still attached to
+   logical node 0, it's just a different physical node. So we have an O(tenant_count) communication
+   to the pageserver, but not an O(tenant_count) write to the database.
+4. S3 writes continue from nodes 0a and 0b, but the writes do not collide due to different
+   node generation in the suffix, and the writes from node 0a are not visible to the rest
+   of the system because endpoints are reading only from node 0b.
```

Comment:
Maybe repeat this example with the k8s StatefulSet to justify the node generation ID more. This replacement node use case here is something we're unlikely to have in practice, I think. The k8s StatefulSet is something I can see happening.

Reply:
In 9d8e3cac3eba024718f01fa3bb450fa7b7a0a525 I've explicitly called out that this node_id reuse is what we would expect in a StatefulSet